### ISSUE TITLE
Questions about VAE training

### ISSUE THREAD

**Issue by LucFrachon** [mkusner/grammarVAE#26]
Hi Matt,
Great piece of work and great to see that you seem to answer questions posted here!
I came across your paper a few months ago and ported it to Pytorch with the intention of using the same principles in a different task. However, I am struggling to get good reconstruction accuracy when training the VAE. The main difference with your training method is that instead of using pre-exisiting data, I generate random sentences on the fly while training (I don't have the same restrictions as SMILES strings over semantic validity). To ensure unbiased training, my sentence lengths are uniform random.

I measure reconstruction accuracy as the proportion of strings (from a randomly generated set of 1000) that are exactly recovered by the model. I am getting 0-5% at best, and the reconstruction loss (BCE) plateaus at a fairly high level. I am annealing the KLD from 0 to 1/latent _size, but there is only so much this achieves.

A few questions:
- What accuracy were you getting, using the same metric (if you did use it)? Did you observe the same behaviour?
- In your experience, was exact reconstruction accuracy essential to the quality of the logP predictor? Maybe it is not such a useful metric in assessing the quality of the latent representations learned by the model, in terms of usability by the BO predictor?
- According to the paper, you were able to train the VAE independently from the logP prediction task, i.e. you didn't need to manipulate the latent space to include relevant  'features' for the BO task. Did you run any experiments where you tried to link the two during the VAE training phase? Do you have any reflections on this?
- In molecule_vae.py, the method _sample_using_masks() uses a Gumbel-Softmax distribution. If my understanding is correct, the main benefit of taking the argmax of a  G-S distribution is that it provides a differentiable alternative to sampling from a classical Softmax (as described in your paper GANs for Sequences of Discrete Elements with the Gumbel-Softmax Distribution). However, as far as I can tell, we never actually backpropagate through this portion of the code, so what was the motivation for using GS rather than Softmax?

Thanks

**Comment by mkusner** [comment]
To your questions:


> What accuracy were you getting, using the same metric (if you did use it)? Did you observe the same behaviour?

So for whatever reason the keras metric that records accuracy is buggy so we don't get reliable accuracy estimates during training. At the end of training we get 53.7% reconstruction accuracy on a particular sample, check out Table 6 and Appendix C here: https://arxiv.org/pdf/1703.01925.pdf


> In your experience, was exact reconstruction accuracy essential to the quality of the logP predictor? Maybe it is not such a useful metric in assessing the quality of the latent representations learned by the model, in terms of usability by the BO predictor?

So I'd say that better reconstruction accuracy certainly improves BO, but it definitely doesn't need to be perfect to get interesting results.

> According to the paper, you were able to train the VAE independently from the logP prediction task, i.e. you didn't need to manipulate the latent space to include relevant 'features' for the BO task. Did you run any experiments where you tried to link the two during the VAE training phase? Do you have any reflections on this?

Sorry we actually don't have results on simultaneously training the VAE and logP. But it would definitely help. Later follow-up papers did this (e.g., https://papers.nips.cc/paper/8005-constrained-graph-variational-autoencoders-for-molecule-design.pdf) and I'm sure it would help this model.

> In molecule_vae.py, the method _sample_using_masks() uses a Gumbel-Softmax distribution. If my understanding is correct, the main benefit of taking the argmax of a G-S distribution is that it provides a differentiable alternative to sampling from a classical Softmax (as described in your paper GANs for Sequences of Discrete Elements with the Gumbel-Softmax Distribution). However, as far as I can tell, we never actually backpropagate through this portion of the code, so what was the motivation for using GS rather than Softmax?

So we don't do Gumbel-Softmax here, we do Gumbel-Argmax, which is equivalent to sampling from a categorical distribution: https://lips.cs.princeton.edu/the-gumbel-max-trick-for-discrete-distributions/  We do this just because (a) it avoids computing exponents which can make the computation numerically unstable, (b) at the time I didn't know of a fast way to sample from that categorical distribution (although I may easily have missed some built in numpy functions).

Finally have you seen these other Pytorch implementations?:
- https://github.com/episodeyang/grammar_variational_autoencoder
- https://github.com/daandouwe/grammar-vae

**Comment by LucFrachon** [comment]
Thanks, really useful answers.
Yes, I saw these Pytorch implementations, mine is very similar obviously but implementing the paper myself helped me understand it!
